server/infinispan/src/main/java/org/hawkular/apm/server/infinispan/InfinispanPropertyCache.java

```diff
@@ -0,0 +1,186 @@
+        this();
+        this.propertyCache = cache;
+        this.propertyCache.addListener(eventListener);
+    }
+
+    @Override
+    @PostConstruct
+    public void init() {
+        // If cache container not already provisions, then must be running outside of a JEE
+        // environment, so create a default cache container
+        if (cacheContainer == null) {
+            log.fine("Using default cache");
+            propertyCache = InfinispanCacheManager.getDefaultCache(CACHE_NAME);
+        } else {
+            log.fine("Using container provided cache");
+            propertyCache = cacheContainer.getCache(CACHE_NAME);
+        }
+
+        propertyCache.addListener(eventListener);
+    }
+
+    @Override
+    public Property get(String tenantId, String id) {
+        Property<?> property = propertyCache.get(id);
+
+        if (property == null) {
+            property = getFromSystemOrEnvVariables(id);
+            if (property == null) {
+                return null;
+            }
+
+            try {
+                store(tenantId, property);
+            } catch (CacheException e) {
+                log.severe(String.format("Could not store environmental property: %s to cache", property));
+            }
+        }
+
+        if (log.isLoggable(Level.FINEST)) {
+            log.finest("Get property [id=" + id + "] = " + property);
+        }
+
+        return property;
+    }
+
+    @Override
+    public void store(String tenantId, List<Property<?>> properties) throws CacheException {
+        if (cacheContainer != null) {
+            propertyCache.startBatch();
+        }
+
+        for (Property<?> property: properties) {
+            if (log.isLoggable(Level.FINEST)) {
+                log.finest("Store property [id=" + property.getId() + "]: " + property);
+            }
+
+            propertyCache.put(property.getId(), property);
+        }
+
+        if (cacheContainer != null) {
+            propertyCache.endBatch(true);
+        }
+    }
+
+    @Override
+    public void store(String tenantId, Property<?> property) throws CacheException {
+        if (log.isLoggable(Level.FINEST)) {
+            log.finest("Store property [id=" + property.getId() + "]: " + property);
+        }
+
+        propertyCache.put(property.getId(), property);
+    }
+
+    @Override
+    public void addObserver(String tenantId, String id, Observer observer) {
+        List<Observer> observers = eventListener.observersMap.get(id);
+        if (observers == null) {
+            observers = new CopyOnWriteArrayList<>();
+            eventListener.observersMap.put(id, observers);
+        }
+
+        observers.add(observer);
+    }
+
+    @Override
+    public void removeObserver(String tenantId, String id, Observer observer) {
+        List<Observer> observers = eventListener.observersMap.get(id);
+        if (observers != null) {
+            observers.remove(observer);
+        }
```

Comment:
Might want to check if observers is empty and if so remove it from the map - just to tidy up.